File: MarkdownEditor/MarkdownEditor/Utilities/DiffEngine.swift
import Foundation

/// Myers diff algorithm for computing line-level diffs between two strings.
enum DiffEngine {

    /// Compute a FileDiff between original and modified text.
    /// - Parameters:
    ///   - original: The original file content
    ///   - modified: The modified file content
    ///   - path: File path for display
    ///   - contextLines: Number of context lines around each change (default 3)
    /// - Returns: A FileDiff with hunks, additions, and deletions
    static func diff(original: String, modified: String, path: String, contextLines: Int = 3) -> FileDiff {
        let oldLines = original.components(separatedBy: "\n")
        let newLines = modified.components(separatedBy: "\n")

        let editScript = computeEditScript(old: oldLines, new: newLines)
        let hunks = buildHunks(editScript: editScript, oldLines: oldLines, newLines: newLines, contextLines: contextLines)

        var totalAdditions = 0
        var totalDeletions = 0
        for hunk in hunks {
            for line in hunk.lines {
                switch line.type {
                case .addition: totalAdditions += 1
                case .deletion: totalDeletions += 1
                case .context: break
                }
            }
        }

        return FileDiff(
            originalPath: path,
            modifiedPath: path,
            hunks: hunks,
            additions: totalAdditions,
            deletions: totalDeletions
        )
    }

    // MARK: - Myers Diff Algorithm

    private enum EditOp {
        case equal(oldIdx: Int, newIdx: Int)
        case insert(newIdx: Int)
        case delete(oldIdx: Int)
    }

    private static func computeEditScript(old: [String], new: [String]) -> [EditOp] {
        let n = old.count
        let m = new.count
        let max = n + m

        if max == 0 { return [] }

        // V array indexed from -max to max; we use offset of max
        var v = Array(repeating: 0, count: 2 * max + 1)
        var trace: [[Int]] = []

        outer: for d in 0...max {
            trace.append(v)
            for k in stride(from: -d, through: d, by: 2) {
                let idx = k + max
                var x: Int
                if d == 0 {
                    x = 0
                } else if k == -d || (k != d && v[idx - 1] < v[idx + 1]) {
                    x = v[idx + 1]
                } else {
                    x = v[idx - 1] + 1
                }
                var y = x - k

                while x < n && y < m && old[x] == new[y] {
                    x += 1
                    y += 1
                }

                v[idx] = x

                if x >= n && y >= m {
                    break outer
                }
            }
        }

        // Backtrack to build edit script
        var ops: [EditOp] = []
        var x = n
        var y = m

        for d in stride(from: trace.count - 1, through: 0, by: -1) {
            let v = trace[d]
            let k = x - y

            var prevK: Int
            if d == 0 {
                // At d=0, we only had k=0, just add remaining equals
                break
            } else if k == -d || (k != d && v[k - 1 + max] < v[k + 1 + max]) {
                prevK = k + 1
            } else {
                prevK = k - 1
            }

            let prevX = v[prevK + max]
            let prevY = prevX - prevK

            // Diagonal (equals)
            while x > prevX && y > prevY {
                x -= 1
                y -= 1
                ops.append(.equal(oldIdx: x, newIdx: y))
            }

            if d > 0 {
                if x == prevX {
                    // Insert
                    y -= 1
                    ops.append(.insert(newIdx: y))
                } else {
                    // Delete
                    x -= 1
                    ops.append(.delete(oldIdx: x))
                }
            }
        }

        // Handle remaining equal lines at the start
        while x > 0 && y > 0 {
            x -= 1
            y -= 1
            ops.append(.equal(oldIdx: x, newIdx: y))
        }

        return ops.reversed()
    }

    // MARK: - Hunk Building

    private static func buildHunks(editScript: [EditOp], oldLines: [String], newLines: [String], contextLines: Int) -> [DiffHunk] {
        // Find ranges of changes
        var changeIndices: [Int] = []
        for (i, op) in editScript.enumerated() {
            switch op {
            case .insert, .delete:
                changeIndices.append(i)
            case .equal:
                break
            }
        }

        if changeIndices.isEmpty { return [] }

        // Group changes that are within contextLines*2 of each other
        var groups: [[Int]] = []
        var currentGroup: [Int] = [changeIndices[0]]

        for i in 1..<changeIndices.count {
            let gap = changeIndices[i] - changeIndices[i - 1]
            if gap <= contextLines * 2 + 1 {
                currentGroup.append(changeIndices[i])
            } else {
                groups.append(currentGroup)
                currentGroup = [changeIndices[i]]
            }
        }
        groups.append(currentGroup)

        // Build hunks from groups
        var hunks: [DiffHunk] = []

        for group in groups {
            let firstChange = group.first!
            let lastChange = group.last!

            let startIdx = Swift.max(0, firstChange - contextLines)
            let endIdx = Swift.min(editScript.count - 1, lastChange + contextLines)

            var lines: [DiffLine] = []
            var oldLineNum = 0
            var newLineNum = 0

            // Calculate starting line numbers
            for i in 0..<startIdx {
                switch editScript[i] {
                case .equal:
                    oldLineNum += 1
                    newLineNum += 1
                case .delete:
                    oldLineNum += 1
                case .insert:
                    newLineNum += 1
                }
            }

            let hunkOldStart = oldLineNum + 1
            let hunkNewStart = newLineNum + 1

            for i in startIdx...endIdx {
                switch editScript[i] {
                case .equal(let oldIdx, _):
                    oldLineNum += 1
                    newLineNum += 1
                    lines.append(DiffLine(
                        type: .context,
                        content: oldLines[oldIdx],
                        oldLineNumber: oldLineNum,
                        newLineNumber: newLineNum
                    ))
                case .delete(let oldIdx):
                    oldLineNum += 1
                    lines.append(DiffLine(
                        type: .deletion,
                        content: oldLines[oldIdx],
                        oldLineNumber: oldLineNum,
                        newLineNumber: nil
                    ))
                case .insert(let newIdx):
                    newLineNum += 1
                    lines.append(DiffLine(
                        type: .addition,
                        content: newLines[newIdx],
                        oldLineNumber: nil,
                        newLineNumber: newLineNum
                    ))
                }
            }

            let hunkOldCount = oldLineNum - hunkOldStart + 1
            let hunkNewCount = newLineNum - hunkNewStart + 1

            hunks.append(DiffHunk(
                oldStart: hunkOldStart,
                oldCount: hunkOldCount,
                newStart: hunkNewStart,
                newCount: hunkNewCount,
                lines: lines
            ))
        }

        return hunks
    }
}

File: android/lib/services/diff_engine.dart
import '../models/git_status.dart';

/// Myers diff algorithm — ported from Swift DiffEngine.
class DiffEngine {
  static FileDiff diff({
    required String original,
    required String modified,
    required String path,
    int contextLines = 3,
  }) {
    final oldLines = original.split('\n');
    final newLines = modified.split('\n');

    final editScript = _computeEditScript(oldLines, newLines);
    final hunks = _buildHunks(
      editScript: editScript,
      oldLines: oldLines,
      newLines: newLines,
      contextLines: contextLines,
    );

    var totalAdditions = 0;
    var totalDeletions = 0;
    for (final hunk in hunks) {
      for (final line in hunk.lines) {
        if (line.type == DiffLineType.addition) totalAdditions++;
        if (line.type == DiffLineType.deletion) totalDeletions++;
      }
    }

    return FileDiff(
      originalPath: path,
      modifiedPath: path,
      hunks: hunks,
      additions: totalAdditions,
      deletions: totalDeletions,
    );
  }

  // MARK: - Myers Diff

  static List<_EditOp> _computeEditScript(List<String> old, List<String> newL) {
    final n = old.length;
    final m = newL.length;
    final maxD = n + m;

    if (maxD == 0) return [];

    // V indexed from -maxD to maxD, offset by maxD
    final v = List<int>.filled(2 * maxD + 1, 0);
    final trace = <List<int>>[];

    outer:
    for (var d = 0; d <= maxD; d++) {
      trace.add(List<int>.from(v));
      for (var k = -d; k <= d; k += 2) {
        final idx = k + maxD;
        int x;
        if (d == 0) {
          x = 0;
        } else if (k == -d || (k != d && v[idx - 1] < v[idx + 1])) {
          x = v[idx + 1];
        } else {
          x = v[idx - 1] + 1;
        }
        var y = x - k;

        while (x < n && y < m && old[x] == newL[y]) {
          x++;
          y++;
        }

        v[idx] = x;

        if (x >= n && y >= m) break outer;
      }
    }

    // Backtrack
    final ops = <_EditOp>[];
    var x = n;
    var y = m;

    for (var d = trace.length - 1; d >= 0; d--) {
      final tv = trace[d];
      final k = x - y;

      int prevK;
      if (d == 0) break;
      if (k == -d || (k != d && tv[k - 1 + maxD] < tv[k + 1 + maxD])) {
        prevK = k + 1;
      } else {
        prevK = k - 1;
      }

      final prevX = tv[prevK + maxD];
      final prevY = prevX - prevK;

      while (x > prevX && y > prevY) {
        x--;
        y--;
        ops.add(_EditOpEqual(x, y));
      }

      if (d > 0) {
        if (x == prevX) {
          y--;
          ops.add(_EditOpInsert(y));
        } else {
          x--;
          ops.add(_EditOpDelete(x));
        }
      }
    }

    while (x > 0 && y > 0) {
      x--;
      y--;
      ops.add(_EditOpEqual(x, y));
    }

    return ops.reversed.toList();
  }

  static List<DiffHunk> _buildHunks({
    required List<_EditOp> editScript,
    required List<String> oldLines,
    required List<String> newLines,
    required int contextLines,
  }) {
    final changeIndices = <int>[];
    for (var i = 0; i < editScript.length; i++) {
      final op = editScript[i];
      if (op is _EditOpInsert || op is _EditOpDelete) {
        changeIndices.add(i);
      }
    }

    if (changeIndices.isEmpty) return [];

    // Group nearby changes
    final groups = <List<int>>[];
    var currentGroup = [changeIndices[0]];

    for (var i = 1; i < changeIndices.length; i++) {
      if (changeIndices[i] - changeIndices[i - 1] <= contextLines * 2 + 1) {
        currentGroup.add(changeIndices[i]);
      } else {
        groups.add(currentGroup);
        currentGroup = [changeIndices[i]];
      }
    }
    groups.add(currentGroup);

    final hunks = <DiffHunk>[];

    for (final group in groups) {
      final firstChange = group.first;
      final lastChange = group.last;

      final startIdx = (firstChange - contextLines).clamp(0, editScript.length - 1);
      final endIdx = (lastChange + contextLines).clamp(0, editScript.length - 1);

      final lines = <DiffLine>[];
      var oldLineNum = 0;
      var newLineNum = 0;

      for (var i = 0; i < startIdx; i++) {
        final op = editScript[i];
        if (op is _EditOpEqual) {
          oldLineNum++;
          newLineNum++;
        } else if (op is _EditOpDelete) {
          oldLineNum++;
        } else if (op is _EditOpInsert) {
          newLineNum++;
        }
      }

      final hunkOldStart = oldLineNum + 1;
      final hunkNewStart = newLineNum + 1;

      for (var i = startIdx; i <= endIdx; i++) {
        final op = editScript[i];
        if (op is _EditOpEqual) {
          oldLineNum++;
          newLineNum++;
          lines.add(DiffLine(
            type: DiffLineType.context,
            content: oldLines[op.oldIdx],
            oldLineNumber: oldLineNum,
            newLineNumber: newLineNum,
          ));
        } else if (op is _EditOpDelete) {
          oldLineNum++;
          lines.add(DiffLine(
            type: DiffLineType.deletion,
            content: oldLines[op.oldIdx],
            oldLineNumber: oldLineNum,
          ));
        } else if (op is _EditOpInsert) {
          newLineNum++;
          lines.add(DiffLine(
            type: DiffLineType.addition,
            content: newLines[op.newIdx],
            newLineNumber: newLineNum,
          ));
        }
      }

      hunks.add(DiffHunk(
        oldStart: hunkOldStart,
        oldCount: oldLineNum - hunkOldStart + 1,
        newStart: hunkNewStart,
        newCount: newLineNum - hunkNewStart + 1,
        lines: lines,
      ));
    }

    return hunks;
  }
}

// MARK: - Internal edit op types

sealed class _EditOp {}

class _EditOpEqual extends _EditOp {
  final int oldIdx;
  final int newIdx;
  _EditOpEqual(this.oldIdx, this.newIdx);
}

class _EditOpInsert extends _EditOp {
  final int newIdx;
  _EditOpInsert(this.newIdx);
}

class _EditOpDelete extends _EditOp {
  final int oldIdx;
  _EditOpDelete(this.oldIdx);
}
